pkg/controller/seed-controller-manager/addon/addon_controller.go

```diff
@@ -532,6 +558,14 @@ func (r *Reconciler) ensureIsInstalled(ctx context.Context, log *zap.SugaredLogg
 
 	cmdLog := log.With("cmd", strings.Join(cmd.Args, " "))
 
+	if addon.Name == "csi" && cluster.Spec.Cloud.Hetzner != nil && cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] {
+		// Between v2.22 and v2.23, there was a change to hetzner CSI driver immutable field fsGroupPolicy
+		// as a result, the CSDriver resource has to be redeployed
+		// https://github.com/kubermatic/kubermatic/issues/12429
+		if err := r.migrateHetznerCSIDriver(ctx, log, cluster); err != nil {
+			return fmt.Errorf("failed to migrate CSI Driver: %w", err)
+		}
```

Comment:
While #10821 did the migration in a different place, I would vastly prefer if this migration logic also set a similar annotation, something like `kubermatic.k8c.io/migrated-hetzner-csi-addon` and checked for it. Running the `migrateHetznerCSIDriver` function every time the CSI driver gets reconciled feels expensive, especially given the fact that I don't think the called client is a cached one (I might be wrong here though).

Reply:
That's a fair point, but I would still better sleep at night if we made sure migration code only actually runs once. We just discovered a case where migration logic continuously ran.

Reply:
ptal @embik, I added an annotation with `Update`. Errors in applying the annotation are purposefully just logged, in the worst case, the annotation will be applied in the next reconciliation. But overall the amortized cost of this will be dominated by the `kubectl apply` behaviour of the addons controller.

Reply:
Thanks!